src/compiler/parser.ts

```diff
@@ -2919,7 +2919,7 @@ namespace ts {
                 case SyntaxKind.QuestionToken:
                     return parseJSDocUnknownOrNullableType();
                 case SyntaxKind.FunctionKeyword:
-                    return parseJSDocFunctionType();
+                    return contextFlags & NodeFlags.JSDoc ? parseJSDocFunctionType() : parseTypeReference();

```

Comment:
IMO, rather than falling back to a `parseTypeReference` here, I think we'd rather have `isStartOfType`'s `SyntaxKind.FunctionKeyword` case be predicated on the context and not have `function` be the start of a type outside of a `JSDoc` context. Maybe.

Reply:
Sounds reasonable, but it turns out that `isStartOfType` is never hit in this path. Here's the relevant call stack:

![Pasted Graphic](https://user-images.githubusercontent.com/3277153/55368305-1713f900-54a6-11e9-83b6-4b0578ee9167.png)

Reply:
Another thing we could potentially try to work with is that `tryParseParenthesizedArrowFunctionExpression` only succeeds because it generously ignores a missing `=>` and goes on to parse the arrow function body. If we disable that looseness in this special case, everything after the `?` wouldn't be interpreted as an arrow function, and I bet the next path the parser tries will be correct.